numba/tests/test_analysis.py

```diff
@@ -551,3 +558,55 @@ def impl(array, x=None, a=None):
                            types.float64, types.NoneType('none')),
                           [None, None],
                           np.zeros((2, 3)), 1.2, None)
+
+
+class TestBranchPrunePostSemanticConstRewrites(TestBranchPruneBase):
+    # Tests that semantic constants rewriting works by virtue of branch pruning
+
+    def test_array_ndim_attr(self):
+
+        def impl(array):
+            if array.ndim == 2:
+                if array.shape[1] == 2:
+                    return 1
+            else:
+                return 10
+
+        self.assert_prune(impl, (types.Array(types.float64, 2, 'C'),), [False,
+                                                                        None],
+                          np.zeros((2, 3)))
+        self.assert_prune(impl, (types.Array(types.float64, 1, 'C'),), [True,
+                                                                        'both'],
+                          np.zeros((2,)))
```

Comment:
Can you add another case that checks that pruning doesn't happen when a non-array type is used?

i.e.
```python
from numba import njit
import collections
import numpy as np

FakeArray = collections.namedtuple('FakeArray', ['ndim'])

fa = FakeArray(ndim=2)

@njit
def foo(fa):
    if fa.ndim == 2:
        return fa.ndim
    else:
        object()
       
    
print(foo(np.ones((2,2))))
foo(fa)  # <--- this fails
```

Reply:
I think I've addressed this now in 2aa5e3e, please let me know what you think, thanks!